### PR TITLE
Pipelines support testing prs from forks

### PR DESCRIPTION
**Pull request by scottwittenburg**
Changes to support running pipelines on pull requests from forks.  At a high level, this change supports:

- Running pipelines with no secrets whatsoever, or whatever secrets do happen to be available 
- A code path for PRs allowing pushing binaries to an alternate mirror.  This will improve developer experience in cases when the first push to a PR branch causes a rebuild to an expensive package, but subsequent pushes wouldn't cause subsequent rebuilds.
- Makes use of #19209 so that only a single install pass is required to install packages (spack would previously have installed a binary cache even if it's `full_hash` was different than the locally computed one).  Allows us to do away with:
  * Do a first pass install with `--only dependencies` and `--cache-only`
  * Replace the re-written `spack.yaml` which was rendered useless by the first install
  * Do a second pass install with `--only package` and `--no-cache`

The changes here are based on #19209 and require that to be merged first.

### CODE REVIEW

**Comment by scottwittenburg** [comment]
@scheibelp Depending on how testing goes the next day or so, I may need to make a few changes, but I think this is ready for at least a first round of review.

File: lib/spack/spack/binary_distribution.py [hunk 1]
```diff
@@ -703,10 +703,23 @@ def generate_package_index(cache_prefix):
...
-    file_list = (
-        entry
-        for entry in web_util.list_url(cache_prefix)
-        if entry.endswith('.yaml'))
+    try:
+        file_list = (
+            entry
+            for entry in web_util.list_url(cache_prefix)
+            if entry.endswith('.yaml'))
+    except KeyError as inst:
```
Comment (opadron) [inline lib/spack/spack/binary_distribution.py:706]: Is this the same `KeyError` that we talked about from the s3 implementation of `list_url`?  I thought we had addressed this error?

File: lib/spack/spack/binary_distribution.py [hunk 2]
```diff
@@ -768,10 +781,23 @@ def generate_key_index(key_prefix, tmpdir=None):
...
-    fingerprints = (
-        entry[:-4]
-        for entry in web_util.list_url(key_prefix, recursive=False)
-        if entry.endswith('.pub'))
+    try:
+        fingerprints = (
+            entry[:-4]
+            for entry in web_util.list_url(key_prefix, recursive=False)
+            if entry.endswith('.pub'))
+    except KeyError as inst:
```
Comment (opadron) [inline lib/spack/spack/binary_distribution.py:784]: Same question about this `KeyError`.

File: lib/spack/spack/cmd/ci.py [hunk 3]
```diff
@@ -304,82 +347,63 @@ def ci_rebuild(args):
...
-            # mirror" (only done if cash reporting is enabled)
+                    env, job_spec, job_spec_yaml_path, buildcache_mirror_url,
+                    cdash_build_id, sign_binaries)
+            except Exception as inst:
+                # If the mirror we're pushing to is on S3 and there's some
+                # permissions problem, for example, we can't just target
+                # that exception type here, since users of the
+                # `spack ci rebuild' may not need or want any dependency
+                # on boto3.  So we use the first non-boto exception type
+                # in the heirarchy:
```
Comment (opadron) [inline lib/spack/spack/cmd/ci.py:387]: I wonder if it makes sense to shim out the few Exception classes needed from boto?  I'll leave it up to you to decide if the maintenance cost is worth it.

**Review by opadron — CHANGES_REQUESTED** [review]
Wondering what the situation is with those `KeyErrors`.  Otherwise, LGTM.

File: lib/spack/spack/binary_distribution.py (hunk 1 above)
Comment (scottwittenburg) [reply lib/spack/spack/binary_distribution.py:706]: It's only addressed here, as far as I know, unless you got a fix merged at some point?

File: lib/spack/spack/cmd/ci.py (hunk 3 above)
Comment (scottwittenburg) [reply lib/spack/spack/cmd/ci.py:387]: I'm not sure it's worth the maintenance cost, but I could be convinced, maybe I'm just not seeing the value.

**Review by opadron — APPROVED** [review]
With respect to the `KeyError`, I'm pretty sure that the issue is on/around this line:

https://github.com/spack/spack/blob/6021ac00af6e4adc8a520cbb4b0f53330e7f4176/lib/spack/spack/util/web.py#L275

If you'd like, feel free to take a stab at moving your fix for this from catching the `KeyError` in your code to that part of the code.

Alternatively, if you could make a new issue with a link to this comment, and assign it to me, I'll fix it after the fact (eventually...).

**Comment by scottwittenburg** [comment]
@scheibelp IIUC @tgamblin hopes to merge this by Wednesday this week.  Do you think you'll have time to review it sometime today?  I'll try to respond to your review suggestions immediately.  Let me know if you want to sync up verbally for some background.

**Review by tgamblin — APPROVED** [review]
@scottwittenburg: this looks good for the moment but we should get to the bottom of the Exceptions that we're squashing.  The error messages should help with that.

**Comment by alalazo** [comment]
@tgamblin Since this PR has been merged I'm not seeing any notification in `develop` for the e4s pipeline. Is this expected?

Here's the last commit that shows it:
Screenshot from 2020-11-18 14-11-36

**Comment by scottwittenburg** [comment]
@alalazo We are in the process of "flipping some switches" in gitlab to support running PRs from forks, and while we're ironing out the wrinkles, I have disabled gitlab pipelines.  I expect we'll have them running again in the next day or so.

**Comment by alalazo** [comment]
@scottwittenburg Thank you for the info. Looking forward to have gitlab pipelines in PRs!

**Pull request merged**